server.go

```diff
@@ -32,39 +36,126 @@ const (
+	azureResourceURL = "https://management.azure.com/"
+	vaultResourceURL = "https://vault.azure.net"
 )
 
+// KMSServiceServer is a gRPC server.
 type KMSServiceServer struct {
 	pathToUnixSocket string
 	net.Listener
 	*grpc.Server
 }
 
+// New creates an instance of the KMS Service Server.
 func New(pathToUnixSocketFile string) *KMSServiceServer {
 	kmsServiceServer := new(KMSServiceServer)
 	kmsServiceServer.pathToUnixSocket = pathToUnixSocketFile
 	fmt.Println(kmsServiceServer.pathToUnixSocket)
 	return kmsServiceServer
 }
 
-func getKeysClient() (kv.ManagementClient, string, string, string, error)  {
-	vmClient := kv.New()
-	token, err := GetKeyvaultToken(AuthGrantType(), configFilePath)
+func getKey() (kv.ManagementClient, string, string, string, error)  {
+	kvClient := kv.New()
+
+	vaultURL, err := getVault()
 	if err != nil {
-		return vmClient, "", "", "", fmt.Errorf("failed to get token, error: %v", err)
+		return kvClient, "", "", "", fmt.Errorf("failed to get vault, error: %v", err)
 	}
 	
-	vmClient.Authorizer = token
-	vaultBaseURL, keyName, keyVersion, err := GetKMSProvider(configFilePath)
+	token, err := GetKeyvaultToken(AuthGrantType(), configFilePath, vaultResourceURL)
 	if err != nil {
-		return vmClient, "", "", "", fmt.Errorf("failed to get KMS provider configs, error: %v", err)
+		return kvClient, "", "", "", fmt.Errorf("failed to get token, error: %v", err)
+	}
+	
+	kvClient.Authorizer = token
+	_, keyName, keyVersion, err := GetKMSProvider(configFilePath)
+	if keyName == nil {
+		return kvClient, "", "", "", fmt.Errorf("failed to find KMS providerKeyName in config")
+	}
+	if keyVersion == nil {
+		fmt.Println("Unable to find KMS providerKeyVersion in configs, getting keyVersion")
+		keyVersion, err := getKeyVersion(kvClient, *vaultURL, *keyName)
+		if err != nil {
+			return kvClient, "", "", "", fmt.Errorf("failed to get/create key, error: %v", err)
+		}
+		version := to.String(keyVersion)
+		index := strings.LastIndex(version, "/" )
+		if (index > -1 && index < len(version)-1) {
+			version = version[index+1:]
+			fmt.Println(version)
+		}
+		return kvClient, *vaultURL, *keyName, version, nil
 	}
-	return vmClient, *vaultBaseURL, *keyName, *keyVersion, nil
+
+	fmt.Println("Verify key version from key vault ", *keyName, *keyVersion, *vaultURL)
+	_, err = kvClient.GetKey(*vaultURL, *keyName, *keyVersion)
+	if err != nil {
+		return kvClient, "", "", "", fmt.Errorf("failed to verify key version, error: %v", err)
+	}
+	
+	return kvClient, *vaultURL, *keyName, *keyVersion, nil
+}
+
+func getVaultsClient() kvmgmt.VaultsClient {
+	subscriptionID, _ := GetSubscriptionId(configFilePath)
+	vaultsClient := kvmgmt.NewVaultsClient(*subscriptionID)
+	token, _ := GetKeyvaultToken(AuthGrantType(), configFilePath, azureResourceURL)
+	vaultsClient.Authorizer = token
+	return vaultsClient
+}
+
+func getVault() (vaultURL *string, err error) {
+	vaultsClient := getVaultsClient()
+	providerVaultName, _, _, err := GetKMSProvider(configFilePath)
```

Comment:
we can safety cache the return of `GetKMSProvider`, let us not get it everytime.